tests/unit/s2n_key_update_threads_test.c

```diff
@@ -0,0 +1,264 @@
+#define S2N_TEST_BYTES_TO_SEND    (S2N_DEFAULT_FRAGMENT_LENGTH * S2N_TEST_RECORD_COUNT)
+
+#define S2N_CIPHER_SUITE_WITH_LIMIT(name, source, limit)                 \
+    struct s2n_cipher_suite name = *(source);                            \
+    struct s2n_record_algorithm _##name##_record_alg = *name.record_alg; \
+    _##name##_record_alg.encryption_limit = limit;                       \
+    name.record_alg = &_##name##_record_alg;
+
+S2N_RESULT s2n_set_key_update_request_for_testing(keyupdate_request request);
+
+static void *s2n_send_random_data(void *arg)
+{
+    struct s2n_connection *conn = (struct s2n_connection *) arg;
+
+    uint8_t buffer[S2N_TEST_BUFFER_SIZE] = "hello world";
+
+    size_t bytes_to_send = S2N_TEST_BYTES_TO_SEND;
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    while (bytes_to_send) {
+        int r = s2n_send(conn, buffer, MIN(sizeof(buffer), bytes_to_send), &blocked);
+        if (r >= 0) {
+            bytes_to_send -= r;
+        } else if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            fprintf(stderr, "Send error: %s. %s\n", s2n_strerror(s2n_errno, NULL), s2n_strerror_debug(s2n_errno, NULL));
+            return NULL;
+        }
+    }
+    return conn;
+}
+
+static void *s2n_recv_random_data(void *arg)
+{
+    struct s2n_connection *conn = (struct s2n_connection *) arg;
+
+    uint8_t buffer[S2N_TEST_BUFFER_SIZE] = { 0 };
+
+    size_t bytes_to_read = S2N_TEST_BYTES_TO_SEND;
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    while (bytes_to_read) {
+        int r = s2n_recv(conn, buffer, MIN(sizeof(buffer), bytes_to_read), &blocked);
+        if (r >= 0) {
+            bytes_to_read -= r;
+        } else if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            fprintf(stderr, "Recv error: %s. %s\n", s2n_strerror(s2n_errno, NULL), s2n_strerror_debug(s2n_errno, NULL));
+            return NULL;
+        }
+    }
+    return conn;
+}
+
+static S2N_RESULT s2n_send_and_recv_random_data(struct s2n_connection *conn)
+{
+    /*
+     * This test is intended to find concurrency issues when sending and receiving
+     * KeyUpdates, so we need to run the reader and writer in separate threads.
+     */
+
+    pthread_t reader = 0;
+    RESULT_ENSURE_EQ(pthread_create(&reader, NULL, s2n_recv_random_data, (void *) conn), 0);
+
+    pthread_t writer = 0;
+    RESULT_ENSURE_EQ(pthread_create(&writer, NULL, s2n_send_random_data, (void *) conn), 0);
+
+    void *reader_return = NULL;
+    RESULT_ENSURE_EQ(pthread_join(reader, &reader_return), 0);
+    RESULT_ENSURE_REF(reader_return);
+
+    void *writer_return = NULL;
+    RESULT_ENSURE_EQ(pthread_join(writer, &writer_return), 0);
+    RESULT_ENSURE_REF(writer_return);
+
+    RESULT_ENSURE_GT(conn->wire_bytes_out, S2N_TEST_BYTES_TO_SEND);
+    RESULT_ENSURE_GT(conn->wire_bytes_in, S2N_TEST_BYTES_TO_SEND);
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_sanity_check_key_updates_sent(struct s2n_connection *conn)
```

Comment:
I *think* you could expand this check to test that the peer also had a keyupdate. Like if you are the client you could test the server's sequence number and wire_bytes_in. But technically each peer checks their own sequence number anyways, and you wouldn't be able to communicate if those were out of sync. So I don't know if this would actually be useful.

Reply:
I could expand the check, but you're right that it's unnecessary. Also, in practice it would work (since we know both connections have the same max fragment size because we use the default) but theoretically / according to the RFC we know nothing about our peer's max fragment size.